pkg/resource/certificate.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2021 The Cockroach Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach-operator/pkg/security"
+	"github.com/cockroachdb/errors"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// SecretsInterface is a subset of methods from client-go's v1.SecretsInterface
+type SecretsInterface interface {
+	Get(context.Context, string, metav1.GetOptions) (*corev1.Secret, error)
+	Create(context.Context, *corev1.Secret, metav1.CreateOptions) (*corev1.Secret, error)
+}
+
+// CertificateFunc describes a function the generates Certificate objects.
+type CertificateFunc func() (security.Certificate, error)
+
+// FindOrCreateCertificateSecret gets the specified secret or creates it using the result of the creator func. The
+// returned result, when not an error, is a secret of type SecretTypeTLS with the crt and key values set appropriately.
+func FindOrCreateCertificateSecret(ctx context.Context, api SecretsInterface, name string, creator CertificateFunc) (security.Certificate, error) {
```

Comment:
It would be cool if we could use this for generating CRDB certs. I didn't attempt that here, but... 

![](https://i.kym-cdn.com/photos/images/newsfeed/000/117/008/soon_honey_beer_bottle.jpg)